### DISCUSSION TITLE
Why a Provider need a Ref to expose its State?

### DISCUSSION THREAD

**Discussion by XuanTung95** [rrousselGit/riverpod#1102] [Q&A] [answered]
**Describe the solution you'd like**

As I understand, a Provider holds a State, a Ref attaches itself to a Provider to listen to State change.
But why do we need Ref to read the State of Provider?
Can Provider just expose its State so we don't need to pass the Ref around?

**Comment by rrousselGit** [comment, marked as answer]
Providers do not hold state, they are completely immutable. The state is stored somewhere else (ProviderContainer).
Providers are merely a unique key that allows accessing the state.

That's why you can override providers or scope a value for a part of the widget tree

**Reply by XuanTung95** [reply]
Could you explain to me why you store state in ProviderContainer and not Provider?
Is it to create multiple ProviderContainer, each container stores a different state of a single Provider?

**Reply by rrousselGit** [reply]
Yes, among other things such as scoping providers

**Reply by rrousselGit** [reply]
It's also quite important to ensure that interactions with the provider are sane

This restriction is part of what makes Providers a lot more scalable than any other global variable

**Reply by XuanTung95** [reply]
Thanks for your explanation.
